File: Problems 0000-0099/05. Longest Palindromic Substring/solution.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();
        string longest = "";
        for (int i = 0; i < n; i++) {
            pair<int, int> one = expandFromCenter(s, i, i);
            pair<int, int> two = expandFromCenter(s, i, i+1);
            int l1 = one.second - one.first + 1;
            if (l1 > longest.length()) {
                longest = s.substr(one.first, l1);
            }
            int l2 = two.second - two.first + 1;
            if (l2 > longest.length()) {
                longest = s.substr(two.first, l2);
            }
        }
        return longest;
    }

    pair<int, int> expandFromCenter(const string& s, int c1, int c2) {
        int n = s.length();
        while (c1 >= 0 && c2 < n && s[c1] == s[c2]) {
            c1--;
            c2++;
        }
        return {c1+1, c2-1};
    }
};

File: Problems 0000-0099/05. Longest Palindromic Substring/solution.java
class Solution {
    public String longestPalindrome(String s) {
        String res = "";
        for (int i = 0; i < s.length(); i++) {
            String odd = expandFromCenter(s, i, i);
            String even = expandFromCenter(s, i, i + 1);

            if (odd.length() > res.length()) {
                res = odd;
            }
            if (even.length() > res.length()) {
                res = even;
            }
        }
        return res;
    }

    public String expandFromCenter(String s, int c1, int c2) {
        while (c1 >= 0 && c2 < s.length() && s.charAt(c1) == s.charAt(c2)) {
            c1--;
            c2++;
        }
        return s.substring(c1 + 1, c2);
    }
}